File: python/src/resp_bench/engine/benchmark.py
"""Async benchmark engine.

Concurrency model: a single asyncio event loop with **one client per
connection** and **one worker coroutine per connection**, run concurrently.
Each worker awaits one command at a time -- i.e. ``pipeline_depth`` is
effectively 1, the same "one in-flight request per connection" shape the Java
engine's virtual-thread workers produce.

A request-based phase target is a single budget shared across all workers, which
they claim from one request at a time -- matching the Java reference's shared
``AtomicLong`` rather than pre-splitting the target per worker.

Known limits of this model, both deliberate:

* **Single event loop.** Above roughly ``HIGH_CONNECTION_WARN_THRESHOLD``
  connections the loop itself, not the driver, becomes the bottleneck, and the
  loop's queuing delay is attributed to the driver in the reported latency. The
  Java engine faced the same ceiling with a single command-issuing thread (see
  ``docs/ARCHITECTURE.md``) and solved it with multiple issuer threads; this
  engine has no equivalent yet, so it warns above the threshold. Do not compare
  high-connection-count Python results against other engines without accounting
  for this.
* **``pipeline_depth > 1``** (multiple in-flight requests per connection) is not
  implemented; a phase requesting it runs at depth 1 with a warning.
"""

from __future__ import annotations

import asyncio
import logging
import time
from typing import List, Optional

from ..client.benchmark_client import AsyncBenchmarkClient
from ..client.factory import BenchmarkClientFactory
from ..command.command import Command, CommandResult
from ..command.factory import CommandFactory
from ..config.driver_config import DriverConfig
from ..config.phase_config import PhaseConfig
from ..config.workload_config import WorkloadConfig
from ..metrics.collector import MetricsCollector
from ..metrics.ndjson_writer import NdjsonWriter
from .command_selector import CommandSelector
from .key_generator import Counter, KeyGenerator
from .rate_limiter import RateLimiter

logger = logging.getLogger("resp_bench")

# Above this many connections the single event loop, not the driver, tends to set
# the throughput ceiling and its queuing delay shows up as driver latency.
HIGH_CONNECTION_WARN_THRESHOLD = 128

# A worker yields at least this often even when every request completes without
# suspending (a cache hit, or an in-memory driver). Without it one connection can
# monopolise a whole duration-based phase while its peers record nothing.
YIELD_EVERY_N_REQUESTS = 64

# Backoff applied after repeated consecutive failures on one connection, so a
# permanently-failing connection cannot spin at full CPU inflating the error
# count. Capped so a transient blip costs almost nothing.
CONSECUTIVE_FAILURE_BACKOFF_AFTER = 8
MAX_FAILURE_BACKOFF_SECONDS = 0.05


async def _gather_all_or_cancel(coros) -> None:
    """Await all coroutines; on the first failure cancel and drain the rest.

    ``asyncio.gather`` deliberately does NOT cancel siblings when one task
    raises, which would leave workers running against clients the caller has
    already closed. This wrapper cancels them and waits for them to finish
    unwinding before re-raising, so no work escapes the phase that started it.
    """
    tasks = [asyncio.ensure_future(c) for c in coros]
    try:
        await asyncio.gather(*tasks)
    except BaseException:
        for task in tasks:
            if not task.done():
                task.cancel()
        await asyncio.gather(*tasks, return_exceptions=True)
        raise


class BenchmarkEngine:
    def __init__(
        self,
        *,
        host: str,
        port: int,
        driver_config: DriverConfig,
        workload_config: WorkloadConfig,
        metrics_path: str,
        commit_id: Optional[str] = None,
    ) -> None:
        self._host = host
        self._port = port
        self._driver_config = driver_config
        self._workload_config = workload_config
        self._writer = NdjsonWriter(metrics_path)
        self._commit_id = commit_id
        # True if any phase ended in ERROR. The CLI turns this into a non-zero
        # exit code so the matrix runner does not score the cell as a good run.
        self._had_error = False

    @property
    def had_error(self) -> bool:
        return self._had_error

    async def run(self) -> None:
        logger.info("Starting benchmark: %s", self._workload_config.name())
        logger.info(
            "Driver: %s, Server mode: %s",
            self._driver_config.driver_id,
            self._driver_config.mode,
        )
        logger.info("Concurrency: asyncio task-per-connection (one client per connection)")
        logger.info("Server: %s:%s", self._host, self._port)

        await self._setup_metadata()

        for phase in self._workload_config.phases:
            await self._execute_phase(phase)

        logger.info("Benchmark completed")

    async def _setup_metadata(self) -> None:
        try:
            sample = await BenchmarkClientFactory.create_and_connect(
                self._host, self._port, self._driver_config
            )
            self._writer.set_metadata(
                commit_id=self._commit_id,
                driver_id=self._driver_config.driver_id,
                primary_driver_version=sample.driver_version(),
                secondary_driver_id=self._driver_config.secondary_driver_id(),
                secondary_driver_version=sample.secondary_driver_version(),
                driver_details=sample.driver_details(),
            )
            logger.info(
                "Metadata: commit=%s, driver=%s, version=%s, details=%s",
                self._commit_id or "N/A",
                self._driver_config.driver_id,
                sample.driver_version(),
                sample.driver_details() or "{}",
            )
            await sample.close()
        except Exception as exc:  # noqa: BLE001 - metadata is best-effort
            logger.warning("Failed to get driver version for metadata: %s", exc)
            self._writer.set_metadata(
                commit_id=self._commit_id,
                driver_id=self._driver_config.driver_id,
                primary_driver_version="unknown",
                secondary_driver_id=self._driver_config.secondary_driver_id(),
                secondary_driver_version=None,
            )

    async def _execute_phase(self, phase: PhaseConfig) -> None:
        logger.info("=== Starting phase: %s (%s) ===", phase.id, phase.description)

        if phase.effective_pipeline_depth() > 1:
            logger.warning(
                "pipeline_depth=%d requested for phase '%s', but the Python engine "
                "does not yet implement pipelining; running at depth 1. Results are "
                "not comparable to pipelined runs of other engines.",
                phase.pipeline_depth,
                phase.id,
            )

        if phase.connections > HIGH_CONNECTION_WARN_THRESHOLD:
            logger.warning(
                "connections=%d exceeds %d: the single event loop is likely the "
                "bottleneck rather than the driver, and loop queuing delay is "
                "reported as driver latency. Treat these results with care and do "
                "not compare them directly against other engines.",
                phase.connections,
                HIGH_CONNECTION_WARN_THRESHOLD,
            )

        collector = MetricsCollector()
        status = "ERROR"
        failure: Optional[BaseException] = None
        clients: List[AsyncBenchmarkClient] = []
        try:
            # Inside the guarded region: a failure part-way through opening
            # connections (server down, maxclients reached) or an unsupported
            # command name must still close what was opened and emit a row.
            clients = await self._create_clients(phase)
            commands = CommandFactory.create_all(phase.commands)

            if phase.warmup_requests > 0:
                await self._warmup(clients, phase.warmup_requests)

            # Started here so the measured window covers the workload only, not
            # connection setup or warmup.
            collector.start()
            status = await self._run_workload(phase, clients, commands, collector)
        except Exception as exc:  # noqa: BLE001 - recorded below, then re-raised
            # Failures used to escape before anything was written, so the phase
            # produced no row at all. Always emit a row so it stays visible.
            failure = exc
            logger.error("Phase %s failed: %s", phase.id, exc)
        finally:
            # A phase that failed before the workload started still needs real
            # timestamps: nulls would violate the documented schema and the graph
            # scripts would aggregate the row as a 0-RPS data point.
            if collector.start_time is None:
                collector.start()
            collector.stop()
            await self._close_clients(clients)

        if status == "ERROR":
            self._had_error = True

        self._writer.write_phase_results(
            phase_id=phase.id,
            status=status,
            connections=phase.connections,
            collector=collector,
        )
        self._log_phase_summary(phase, collector, status)

        if failure is not None:
            raise failure

    async def _create_clients(self, phase: PhaseConfig) -> List[AsyncBenchmarkClient]:
        logger.info("Creating %d connections...", phase.connections)
        cps_limiter = RateLimiter.create(phase.cps_limit) if phase.has_cps_limit() else None

        clients: List[AsyncBenchmarkClient] = []
        for _ in range(phase.connections):
            if cps_limiter is not None:
                await cps_limiter.acquire()
            client = await BenchmarkClientFactory.create_and_connect(
                self._host, self._port, self._driver_config
            )
            clients.append(client)
        logger.info("All %d connections established", len(clients))
        return clients

    async def _warmup(self, clients: List[AsyncBenchmarkClient], warmup_requests: int) -> None:
        logger.info("Warmup: %d PINGs per client...", warmup_requests)

        async def warm(client: AsyncBenchmarkClient) -> None:
            for _ in range(warmup_requests):
                result = await client.ping()
                # Fail fast on an unreachable/misconfigured server rather than
                # running a whole phase that records only errors.
                if not result.success:
                    raise RuntimeError(f"Warmup PING failed: {result.error}")

        # Cancel-and-drain on first failure so no warmup task keeps pinging a
        # client that _execute_phase has already closed.
        await _gather_all_or_cancel(warm(c) for c in clients)
        logger.info("Warmup completed")

    async def _run_workload(
        self,
        phase: PhaseConfig,
        clients: List[AsyncBenchmarkClient],
        commands: List[Command],
        collector: MetricsCollector,
    ) -> str:
        completion = phase.completion
        num_workers = len(clients)
        seed_base = phase.keyspace.seed_value()
        shared_counter = Counter()  # shared across workers for sequential_int

        # Constructed here, at the start of the measured window, rather than
        # before warmup: the limiter's schedule starts from its construction
        # time, so building it earlier would let the phase open with a burst of
        # accumulated slots proportional to the warmup duration.
        rate_limiter = RateLimiter.create(phase.rps_limit) if phase.has_rps_limit() else None

        # A request-based target is a single budget SHARED across workers, which
        # each worker claims from one request at a time (matching the Java
        # reference's shared AtomicLong). A slow connection therefore cannot cap
        # the phase -- faster workers absorb the slack and the phase ends when
        # the total budget is exhausted. Duration-based runs use a wall-clock
        # deadline instead.
        target_requests = None if completion.is_duration_based() else completion.total_requests()
        end_time = (
            time.monotonic() + completion.duration_seconds()
            if completion.is_duration_based()
            else None
        )
        request_budget = Counter()

        async def worker(idx: int, client: AsyncBenchmarkClient) -> None:
            key_gen = KeyGenerator.create_with_seed(
                phase.keyspace, seed_base + idx, sequential_counter=shared_counter
            )
            selector = CommandSelector(commands)
            since_yield = 0
            consecutive_failures = 0
            while True:
                if target_requests is not None:
                    # Claim a slot; claims are atomic on the single event loop
                    # (no await between read and increment), so unlike Java no
                    # decrement-on-overshoot is needed.
                    if request_budget.next_value() >= target_requests:
                        break
                elif time.monotonic() >= end_time:
                    break

                if rate_limiter is not None:
                    await rate_limiter.acquire()

                command = None
                succeeded = False
                try:
                    command = selector.select()
                    key = key_gen.next_key()
                    result = await command.execute(client, key)
                    collector.record(result)
                    succeeded = result.success
                except Exception:  # noqa: BLE001 - record failures, keep going
                    collector.record(
                        CommandResult(
                            command_name=command.name if command else "UNKNOWN",
                            latency_micros=0,
                            success=False,
                        )
                    )

                # A request can complete without ever suspending -- a driver that
                # raises before its first await (caught by _measure), or any
                # in-memory/cache-hit response. Such a worker would monopolise the
                # event loop, so guarantee a suspension point.
                since_yield += 1
                if succeeded:
                    consecutive_failures = 0
                    if since_yield >= YIELD_EVERY_N_REQUESTS:
                        since_yield = 0
                        await asyncio.sleep(0)
                else:
                    consecutive_failures += 1
                    since_yield = 0
                    if consecutive_failures >= CONSECUTIVE_FAILURE_BACKOFF_AFTER:
                        # Back off so a permanently-failing connection cannot spin
                        # at full CPU inflating the error count.
                        await asyncio.sleep(
                            min(
                                0.001 * (consecutive_failures - CONSECUTIVE_FAILURE_BACKOFF_AFTER + 1),
                                MAX_FAILURE_BACKOFF_SECONDS,
                            )
                        )
                    else:
                        await asyncio.sleep(0)

        logger.info("Starting %d worker coroutines...", num_workers)
        try:
            # Cancel-and-drain on first failure: a worker that dies must not
            # leave its peers running against clients we are about to close.
            await _gather_all_or_cancel(worker(i, c) for i, c in enumerate(clients))
        except KeyboardInterrupt:  # pragma: no cover
            logger.warning("Workload interrupted")
            return "INTERRUPTED"
        except Exception as exc:  # noqa: BLE001
            logger.error("Error during workload execution: %s", exc)
            return "ERROR"

        # A phase in which nothing succeeded produced no usable latency data;
        # reporting COMPLETED would let the orchestrator record it as a good run.
        successes = collector.total_requests - collector.total_errors
        if collector.total_requests > 0 and successes == 0:
            logger.error(
                "All %d requests failed; reporting phase as ERROR",
                collector.total_requests,
            )
            return "ERROR"

        logger.info("All operations completed (%d total requests)", collector.total_requests)
        return "COMPLETED"

    async def _close_clients(self, clients: List[AsyncBenchmarkClient]) -> None:
        logger.info("Closing %d connections...", len(clients))
        for client in clients:
            try:
                await client.close()
            except Exception as exc:  # noqa: BLE001
                logger.warning("Error closing client: %s", exc)

    def _log_phase_summary(self, phase, collector, status) -> None:
        duration_s = collector.duration_millis() / 1000.0
        total = collector.total_requests
        errors = collector.total_errors
        rps = round(total / duration_s) if duration_s > 0 else 0
        logger.info("=== Phase %s completed: %s ===", phase.id, status)
        logger.info(
            "  Duration: %.1fs | Requests: %d | Errors: %d | RPS: %d",
            duration_s,
            total,
            errors,
            rps,
        )
        for cmd_name, cmd_metrics in collector.command_metrics.items():
            if cmd_metrics.count() == 0:
                continue
            logger.info(
                "  %s: %d req (%d err) | p50=%dus p95=%dus p99=%dus p99.9=%dus | min=%dus max=%dus",
                cmd_name,
                cmd_metrics.requests,
                cmd_metrics.errors,
                cmd_metrics.percentile(50),
                cmd_metrics.percentile(95),
                cmd_metrics.percentile(99),
                cmd_metrics.percentile(99.9),
                cmd_metrics.min(),
                cmd_metrics.max(),
            )

File: node/src/engine/benchmark.ts
/**
 * Benchmark engine.
 *
 * Concurrency model: a single event loop with **one client per connection** (the
 * `client == connection` invariant every engine holds) and **one worker per
 * connection**, all started together via `Promise.all`. Node is single-threaded
 * with async I/O, so this is the faithful analogue of Java's
 * virtual-thread-per-client design -- an awaited command parks the worker, not
 * the loop, so the other connections keep making progress.
 *
 * Two loops, mirroring Java (BenchmarkEngine.java:353-452):
 * - `pipeline_depth <= 1`: issue one command, await it, record, repeat.
 * - `pipeline_depth > 1`: keep up to `pipelineDepth` requests in flight per
 *   connection, awaiting whichever settles first and immediately refilling.
 *
 * The request budget is **shared across all workers**, claimed one request at a
 * time, exactly as Java does with its per-phase `AtomicLong`
 * (BenchmarkEngine.java:249, 363-367). It is deliberately not pre-divided per
 * worker: with a shared budget a slow connection cannot cap the run -- faster
 * workers absorb the slack and the phase ends when the total budget is spent.
 * Pre-splitting would bound wall-clock by the slowest connection and change the
 * per-connection distribution, which is a real cross-engine comparability gap.
 */

import type { BenchmarkClient } from '../client/benchmarkClient.js';
import { BenchmarkClientFactory } from '../client/factory.js';
import type { Command, CommandResult } from '../command/command.js';
import { CommandFactory } from '../command/factory.js';
import type { DriverConfig } from '../config/driverConfig.js';
import type { PhaseConfig } from '../config/phaseConfig.js';
import type { WorkloadConfig } from '../config/workloadConfig.js';
import { MetricsCollector } from '../metrics/collector.js';
import { NdjsonWriter } from '../metrics/ndjsonWriter.js';
import { CommandSelector } from './commandSelector.js';
import { Counter, KeyGenerator } from './keyGenerator.js';
import { RateLimiter } from './rateLimiter.js';

const PROGRESS_LOG_INTERVAL_MS = 10_000;
const CONNECTION_LOG_INTERVAL = 50;

export interface Logger {
  info(message: string): void;
  warn(message: string): void;
  error(message: string): void;
}

export const consoleLogger: Logger = {
  info: (message) => console.log(`${new Date().toISOString()} INFO  ${message}`),
  warn: (message) => console.warn(`${new Date().toISOString()} WARN  ${message}`),
  error: (message) => console.error(`${new Date().toISOString()} ERROR ${message}`),
};

/**
 * A shared, monotonically-drained request budget for one phase.
 *
 * `claim()` is atomic without a lock: it reads and writes with no `await` in
 * between, so concurrent workers on the single event-loop thread can never
 * interleave inside it.
 */
class RequestBudget {
  private remaining: number;

  constructor(total: number) {
    this.remaining = total;
  }

  claim(): boolean {
    if (this.remaining <= 0) return false;
    this.remaining -= 1;
    return true;
  }
}

export interface BenchmarkEngineOptions {
  host: string;
  port: number;
  driverConfig: DriverConfig;
  workloadConfig: WorkloadConfig;
  metricsPath: string;
  commitId?: string | null;
  logger?: Logger;
}

export class BenchmarkEngine {
  private readonly host: string;
  private readonly port: number;
  private readonly driverConfig: DriverConfig;
  private readonly workloadConfig: WorkloadConfig;
  private readonly writer: NdjsonWriter;
  private readonly commitId: string | null;
  private readonly log: Logger;

  constructor(options: BenchmarkEngineOptions) {
    this.host = options.host;
    this.port = options.port;
    this.driverConfig = options.driverConfig;
    this.workloadConfig = options.workloadConfig;
    this.writer = new NdjsonWriter(options.metricsPath);
    this.commitId = options.commitId ?? null;
    this.log = options.logger ?? consoleLogger;
  }

  async run(): Promise<void> {
    this.log.info(`Starting benchmark: ${this.workloadConfig.name()}`);
    this.log.info(`Driver: ${this.driverConfig.driverId}, Server mode: ${this.driverConfig.mode}`);
    this.log.info('Concurrency: event-loop task-per-connection (one client per connection)');
    this.log.info(`Server: ${this.host}:${this.port}`);

    await this.setupMetadata();

    for (const phase of this.workloadConfig.phases) {
      await this.executePhase(phase);
    }

    this.log.info('Benchmark completed');
  }

  /** Best-effort: a version lookup failure must not fail the benchmark. */
  private async setupMetadata(): Promise<void> {
    try {
      const sample = await BenchmarkClientFactory.createAndConnect(
        this.host,
        this.port,
        this.driverConfig,
      );
      const version = sample.driverVersion();
      this.writer.setMetadata({
        commitId: this.commitId,
        driverId: this.driverConfig.driverId,
        primaryDriverVersion: version,
        secondaryDriverId: this.driverConfig.secondaryDriverId(),
        secondaryDriverVersion: sample.secondaryDriverVersion?.() ?? null,
      });
      this.log.info(
        `Metadata: commit=${this.commitId ?? 'N/A'}, driver=${this.driverConfig.driverId}, version=${version}`,
      );
      await sample.close();
    } catch (error) {
      this.log.warn(`Failed to get driver version for metadata: ${(error as Error).message}`);
      this.writer.setMetadata({
        commitId: this.commitId,
        driverId: this.driverConfig.driverId,
        primaryDriverVersion: 'unknown',
        secondaryDriverId: this.driverConfig.secondaryDriverId(),
        secondaryDriverVersion: null,
      });
    }
  }

  private async executePhase(phase: PhaseConfig): Promise<void> {
    this.log.info(`=== Starting phase: ${phase.id} (${phase.description ?? ''}) ===`);

    const collector = new MetricsCollector();
    const clients = await this.createClients(phase);
    const commands = CommandFactory.createAll(phase.commands);
    const rateLimiter = phase.hasRpsLimit() ? RateLimiter.create(phase.rpsLimit) : null;

    let status: string;
    try {
      if (phase.warmupRequests > 0) await this.warmup(clients, phase.warmupRequests);

      collector.start();
      status = await this.runWorkload(phase, clients, commands, rateLimiter, collector);
      collector.stop();
    } finally {
      await this.closeClients(clients);
    }

    this.writer.writePhaseResults({
      phaseId: phase.id,
      status,
      connections: phase.connections,
      collector,
    });
    this.logPhaseSummary(phase, collector, status);
  }

  private async createClients(phase: PhaseConfig): Promise<BenchmarkClient[]> {
    this.log.info(`Creating ${phase.connections} connections...`);
    const cpsLimiter = phase.hasCpsLimit() ? RateLimiter.create(phase.cpsLimit) : null;

    const clients: BenchmarkClient[] = [];
    for (let i = 0; i < phase.connections; i++) {
      if (cpsLimiter !== null) await cpsLimiter.acquire();
      clients.push(
        await BenchmarkClientFactory.createAndConnect(this.host, this.port, this.driverConfig),
      );
      if ((i + 1) % CONNECTION_LOG_INTERVAL === 0) {
        this.log.info(`Created ${i + 1}/${phase.connections} connections`);
      }
    }
    this.log.info(`All ${clients.length} connections established`);
    return clients;
  }

  /**
   * Send warmup PINGs on every client, failing fast if any of them errors.
   *
   * A dead or misconfigured server would otherwise produce a whole phase of
   * nothing but errors, which is far harder to diagnose than an upfront throw.
   */
  private async warmup(clients: BenchmarkClient[], warmupRequests: number): Promise<void> {
    this.log.info(`Warmup: ${warmupRequests} PING(s) per client...`);
    // Warmup mode lets the recording driver suppress simulated errors, so an
    // error_rate workload is not aborted by the very errors it is measuring.
    for (const client of clients) client.setWarmupMode?.(true);
    try {
      await Promise.all(
        clients.map(async (client) => {
          for (let i = 0; i < warmupRequests; i++) {
            const result = await client.ping();
            if (result.error !== undefined) {
              throw new Error(`Warmup PING failed: ${result.error.message}`);
            }
          }
        }),
      );
    } finally {
      for (const client of clients) client.setWarmupMode?.(false);
    }
    this.log.info('Warmup completed');
  }

  private async runWorkload(
    phase: PhaseConfig,
    clients: BenchmarkClient[],
    commands: Command[],
    rateLimiter: RateLimiter | null,
    collector: MetricsCollector,
  ): Promise<string> {
    const { completion } = phase;
    const pipelineDepth = phase.effectivePipelineDepth();
    const seedBase = phase.keyspace.seedValue();
    // Shared across workers for sequential_int, so they collectively emit
    // 0, 1, 2, ... exactly as the Java reference does.
    const sharedCounter = new Counter();

    const budget = completion.isRequestBased() ? new RequestBudget(completion.totalRequests()) : null;
    const deadlineMs = completion.isDurationBased()
      ? Date.now() + completion.durationSeconds() * 1000
      : null;

    const keepGoing = (): boolean => {
      if (deadlineMs !== null && Date.now() >= deadlineMs) return false;
      if (budget !== null) return budget.claim();
      return true;
    };

    this.log.info(
      `Starting ${clients.length} workers (pipeline_depth=${pipelineDepth})...`,
    );

    const progressTimer = setInterval(() => {
      this.logProgress(collector, completion.isRequestBased() ? completion.totalRequests() : null);
    }, PROGRESS_LOG_INTERVAL_MS);
    // Do not let the interval hold the event loop open past the phase.
    progressTimer.unref();

    try {
      await Promise.all(
        clients.map((client, index) => {
          const keyGen = KeyGenerator.createWithSeed(phase.keyspace, seedBase + index, sharedCounter);
          const selector = new CommandSelector(commands);
          return pipelineDepth <= 1
            ? this.runSyncLoop(client, selector, keyGen, rateLimiter, collector, keepGoing)
            : this.runPipelinedLoop(
                client,
                selector,
                keyGen,
                rateLimiter,
                collector,
                keepGoing,
                pipelineDepth,
              );
        }),
      );
      this.log.info(`All operations completed (${collector.totalRequests} total requests)`);
      return 'COMPLETED';
    } catch (error) {
      this.log.error(`Error during workload execution: ${(error as Error).message}`);
      return 'ERROR';
    } finally {
      clearInterval(progressTimer);
    }
  }

  /** One in-flight request per connection (pipeline_depth <= 1). */
  private async runSyncLoop(
    client: BenchmarkClient,
    selector: CommandSelector,
    keyGen: KeyGenerator,
    rateLimiter: RateLimiter | null,
    collector: MetricsCollector,
    keepGoing: () => boolean,
  ): Promise<void> {
    while (keepGoing()) {
      if (rateLimiter !== null) await rateLimiter.acquire();
      collector.record(await this.issue(client, selector, keyGen));
    }
  }

  /**
   * Up to `pipelineDepth` in-flight requests per connection.
   *
   * Each slot runs its own claim/issue/record cycle, so a settled request is
   * replaced immediately rather than waiting on a whole batch -- the same
   * "refill as they land" behaviour as Java's `anyOf` loop, expressed as N
   * independent slot loops sharing the connection.
   */
  private async runPipelinedLoop(
    client: BenchmarkClient,
    selector: CommandSelector,
    keyGen: KeyGenerator,
    rateLimiter: RateLimiter | null,
    collector: MetricsCollector,
    keepGoing: () => boolean,
    pipelineDepth: number,
  ): Promise<void> {
    const slot = async (): Promise<void> => {
      while (keepGoing()) {
        if (rateLimiter !== null) await rateLimiter.acquire();
        collector.record(await this.issue(client, selector, keyGen));
      }
    };
    await Promise.all(Array.from({ length: pipelineDepth }, slot));
  }

  /**
   * Select and run one command.
   *
   * Drivers already convert a rejection into a failed TimedResult, so a throw
   * here means an engine-level bug rather than a server error; record it as a
   * failed request and keep the phase running.
   */
  private async issue(
    client: BenchmarkClient,
    selector: CommandSelector,
    keyGen: KeyGenerator,
  ): Promise<CommandResult> {
    const command = selector.select();
    // Only advance the key sequence for commands that consume a key, so PING
    // does not silently shift the sequence other engines produce.
    const key = command.usesKey ? keyGen.nextKey() : '';
    try {
      return await command.execute(client, key);
    } catch (error) {
      return {
        commandName: command.name,
        latencyMicros: 0,
        success: false,
        errorMessage: (error as Error).message,
      };
    }
  }

  private async closeClients(clients: BenchmarkClient[]): Promise<void> {
    this.log.info(`Closing ${clients.length} connections...`);
    for (const client of clients) {
      try {
        await client.close();
      } catch (error) {
        this.log.warn(`Error closing client: ${(error as Error).message}`);
      }
    }
  }

  private logProgress(collector: MetricsCollector, target: number | null): void {
    const elapsedMs = Date.now() - collector.startTime();
    if (elapsedMs <= 0) return;
    const current = collector.totalRequests;
    const rate = Math.round((current * 1000) / elapsedMs);
    if (target !== null) {
      const percent = ((current * 100) / target).toFixed(1);
      this.log.info(`Progress: ${current}/${target} requests (${percent}%) - ${rate} req/s`);
    } else {
      this.log.info(`Progress: ${current} requests - ${rate} req/s`);
    }
  }

  private logPhaseSummary(phase: PhaseConfig, collector: MetricsCollector, status: string): void {
    const durationSeconds = collector.durationMillis() / 1000;
    const rps = durationSeconds > 0 ? Math.round(collector.totalRequests / durationSeconds) : 0;
    this.log.info(`=== Phase ${phase.id} completed: ${status} ===`);
    this.log.info(
      `  Duration: ${durationSeconds.toFixed(1)}s | Requests: ${collector.totalRequests} | ` +
        `Errors: ${collector.totalErrors} | RPS: ${rps}`,
    );
    for (const [name, m] of collector.commandMetrics) {
      if (m.count() === 0 && m.errors === 0) continue;
      this.log.info(
        `  ${name}: ${m.requests} req (${m.errors} err) | ` +
          `p50=${m.percentile(50)}us p95=${m.percentile(95)}us p99=${m.percentile(99)}us ` +
          `p99.9=${m.percentile(99.9)}us | min=${m.min()}us max=${m.max()}us`,
      );
    }
  }
}
